runtime/domain.c

```diff
@@ -821,15 +885,32 @@ CAMLexport void (*caml_leave_blocking_section_hook)(void) =
    caml_leave_blocking_section_default;
 
 CAMLexport void caml_leave_blocking_section() {
-  caml_plat_lock(&domain_self->roots_lock);
+  dom_internal* self = domain_self;
+
+  atomic_store_rel(&self->backup_thread_msg, MSG_ENTERING_OCAML);
+
+  if (self->backup_thread_running) {
+    /* Lock is necessary here to avoid races with the backup thread
+     * [backup_thread_func]. */
+    caml_plat_lock(&self->interruptor.lock);
+    caml_plat_signal(&self->interruptor.cond);
+    caml_plat_unlock(&self->interruptor.lock);
+  }
+
+  caml_plat_lock(&self->domain_lock);
   caml_leave_blocking_section_hook();
   caml_process_pending_signals();
 }
 
 CAMLexport void caml_enter_blocking_section() {
+  dom_internal* self = domain_self;
+
   caml_process_pending_signals();
   caml_enter_blocking_section_hook();
-  caml_plat_unlock(&domain_self->roots_lock);
+  atomic_store_rel(&self->backup_thread_msg, MSG_IN_BLOCKING_SECTION);
+  /* Wakeup backup thread if it is sleeping */
+  caml_plat_signal(&self->domain_cond);
+  caml_plat_unlock(&self->domain_lock);
```

Comment:
Is it worth having an assert that if `!caml_domain_alone()` then the backup thread is running?

Reply:
Done.